src/categorical_algebra/Diagrams.jl

```diff
@@ -105,48 +107,64 @@ type `DiagramHom{id}` and `DiagramHom{op}` induce morphisms between colimits and
 between limits of the diagrams, respectively, whereas morphisms of type
 `DiagramHom{co}` generalize morphisms of polynomial functors.
 """
-struct DiagramHom{T,C<:Cat,F<:FinFunctor,Φ,D<:Functor{<:FinCat,C}}
+abstract type DiagramHom{T,C<:Cat} end
+
+struct SimpleDiagramHom{T,C<:Cat,F<:FinFunctor,Φ<:FinTransformation,D<:Functor{<:FinCat,C}}<:DiagramHom{T,C}
   shape_map::F
   diagram_map::Φ # bug: Φ should be type constrained to be a FinTransformation  
   precomposed_diagram::D
 end
+struct QueryDiagramHom{T,C<:Cat,F<:FinFunctor,Φ<:FinTransformation,D<:Functor{<:FinCat,C},Params<:AbstractDict}<:DiagramHom{T,C}
+  shape_map::F
+  diagram_map::Φ
+  precomposed_diagram::D
+  params::Params
+end
+
 DiagramHom{T}(shape_map::F, diagram_map::Φ, precomposed_diagram::D) where
     {T,C,F<:FinFunctor,Φ<:FinTransformation,D<:Functor{<:FinCat,C}} =
-  DiagramHom{T,C,F,Φ,D}(shape_map, diagram_map, precomposed_diagram)
-
+  SimpleDiagramHom{T,C,F,Φ,D}(shape_map, diagram_map, precomposed_diagram)
+#makes sure you'll never accidentally build a querydiagramhom by building a diagramhom 
+#with no params.
+DiagramHom{T}(shape_map::F, diagram_map::Φ, precomposed_diagram::D,params::Params) where
+    {T,C,F<:FinFunctor,Φ<:FinTransformation,D<:Functor{<:FinCat,C},Params<:AbstractDict} =
+  isempty(params) ?   
+    SimpleDiagramHom{T,C,F,Φ,D}(shape_map, diagram_map, precomposed_diagram) :
+    QueryDiagramHom{T,C,F,Φ,D,Params}(shape_map, diagram_map, precomposed_diagram,params)
+"""Convert the diagram category in which a diagram hom is being viewed."""
 DiagramHom{T}(f::DiagramHom) where T =
-  DiagramHom{T}(f.shape_map, f.diagram_map, f.precomposed_diagram)
+  DiagramHom{T}(f.shape_map, f.diagram_map, f.precomposed_diagram,params=get_params(f))
 
-DiagramHom{T}(ob_maps, hom_map, D::Diagram{T}, D′::Diagram{T}) where T =
-  DiagramHom{T}(ob_maps, hom_map, diagram(D), diagram(D′))
+DiagramHom{T}(ob_maps, hom_map, D::Diagram{T}, D′::Diagram{T};params=Dict()) where T =
+  DiagramHom{T}(ob_maps, hom_map, diagram(D), diagram(D′);params=params)
```

Comment:
For this kind of keyword delegation, the recommended approach is

```julia
DiagramHom{T}(ob_maps, hom_map, D::Diagram{T}, D′::Diagram{T}; kw...) where T =
  DiagramHom{T}(ob_maps, hom_map, diagram(D), diagram(D′); kw...)
```

which saves on repetition and makes it easier to refactor/add keyword args later.